### PR TITLE
Autoscaler implementation

### PR DESCRIPTION
**Pull request by ramr**
This is step 2 of the autoscaler implementation - building on the types defined by @pweil-  in https://github.com/GoogleCloudPlatform/kubernetes/pull/5492 
and the original autoscaler proposal from @smarterclayton at: #2863

To test, define a replication controller and an autoscaler configuration - example at: https://github.com/ramr/autoscaler-k8s-test/blob/master/test-as.json
and the replication controller gets scaled automatically to the min/max limits set in the autoscaler spec.

Or you can use the test repo at https://github.com/ramr/autoscaler-k8s-test

```
git clone https://github.com/ramr/autoscaler-k8s-test  
cd autoscaler-k8s-test  
make K8S_SOURCE_DIR=~/work/kubernetes test  
```

and the replicas for the sample replication controller `cluster/kubectl.sh  get rc`, should scale upto a limit of 6 replicas.

PTAL - Any comments/feedback greatly appreciated. Thx

### CODE REVIEW

**Comment by ramr** [comment]
Signed the CLA.

**Comment by ramr** [comment]
@pweil-  Think the bot requires you to agree as well - since I used the original types from the unmerged PR - https://github.com/GoogleCloudPlatform/kubernetes/pull/5492

**Comment by pweil-** [comment]
Confirmed.  I am @pweil- and I endorse this PR (politician smile)

**Comment by ramr** [comment]
(bump) + any way to check the failing builds?
Shippable says ```The build you're looking for wasn't found.```.

@smarterclayton FYI

**Comment by smarterclayton** [comment]
ok to test

**Comment by smarterclayton** [comment]
```
 ./hack/verify-description.sh
API file is missing the required field descriptions: ./pkg/api/v1beta3/types.go
API file is missing the required field descriptions: ./pkg/api/v1/types.go
```

**Comment by ramr** [comment]
:sos:   can't see the ```Shippable``` errors ``` The build you're looking for wasn't found.```?
Would be nice to know what tests the shippable stuff is running ...

**Comment by smarterclayton** [comment]
See the shippable yaml file

**Comment by ramr** [comment]
k - third time's a charm!!  PTAL - thx.

File: pkg/api/v1/types.go [hunk 1]
```diff
@@ -1976,3 +1976,136 @@ type RangeAllocation struct {
...
+
+	// Advisors defines a set of advisory sources that the auto-scaler
+	// uses to examine if thresholds are (b)reached.
+	Advisors []string `json:"advisors,omitempty" description:"list of auto-scale advisors"`
+}
+
+// AutoScaleThreshold is a behavior based on statistics used to drive the auto-scaler in scaling decisions.
+type AutoScaleThreshold struct {
+	// Type is the type of threshold being used, intention or value.
+	Type AutoScaleThresholdType `json:"type,omitempty" description:"auto-scale threshold type; intention or value based"`
```
Comment (davidopp) [inline pkg/api/v1/types.go]: it wasn't clear to me how value-based works; intention-based is configured in Intentions field, but where is value-based configured?

Comment (ramr) [inline pkg/api/v1/types.go]: Good question - don't think it was anywhere but am not sure if I have the right answer to that. 

The types were as defined in the original autoscaler types PR  ttps://github.com/GoogleCloudPlatform/kubernetes/pull/5492, and when I started to use them, found a couple of missing configuration pieces to evaluate the thresholds - the value and duration in particular, which I added to the ```AutoScaleIntentionThresholdConfig``` struct. Given what the intention thresholds support now, the value based thresholds could simply be folded into the intention based ones (aka just need the intention based thresholds - value based becomes a subset).

Don't know the history/discussions leading to the initial autoscaler types PR, so maybe @pweil-  or @smarterclayton  can shed more light on that.  @pweil- / @smarterclayton ?

Comment (pweil-) [inline pkg/api/v1/types.go]: As the discussion progressed intention based threshold were the way everyone was leaning.  I eventually removed the value based thresholds from the code (https://github.com/GoogleCloudPlatform/kubernetes/pull/5492#discussion_r26480673).  The type is probably unnecessary at this point unless we intend to add value-based configs.

**Comment by ramr** [comment]
Ok - this has been around for a while, so had to do some babysitting - rebased this to the latest code. 

Can't see the build errors - only see a ```Build cannot be found``` error on shippable. 

Any comments on what else is needed here + can you please let me know the shippable errors. 
I can address the errors + comment  re: autoscale threshold type that @davidopp  had along w/ anything else that needs to be done here.  Thanks.

**Comment by davidopp** [comment]
Unfortunately I don't think we're going to be able to give this attention until 1.0.

**Comment by smarterclayton** [comment]
@pmorie and I will do the bulk of the review for this, but we probably want
whomever has been involved in autoscaling on the Google side to take a look
as well.

On Jul 4, 2015, at 2:29 AM, David Oppenheimer <notifications@github.com>
wrote:

Unfortunately I don't think we're going to be able to give this attention
until 1.0.

—
Reply to this email directly or view it on GitHub
<https://github.com/GoogleCloudPlatform/kubernetes/pull/9612#issuecomment-118468747>
.

**Comment by bgrant0607** [comment]
cc @davidopp

**Comment by smarterclayton** [comment]
As discussed offline, if there are others who are interested in picking up
the autoscaler implementation and run with it for 1.1 we'd be happy to
assist or hand that off.

On Tue, Jul 28, 2015 at 7:08 AM, Brian Grant <notifications@github.com>
wrote:

> cc @davidopp <https://github.com/davidopp>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/9612#issuecomment-125560305>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

**Comment by jszczepkowski** [comment]
The updated design of simplified pod autoscaler is described in #12087. We are planning to implement it as a part of 1.1.

**Comment by ramr** [comment]
@jszczepkowski  yeah saw that afterwards. Rebased as it was outdated - was on the wait queue for a while now (almost a couple of months). Anyway that's a moot point now ... 

@davidopp  want me to close this out?

**Comment by ncdc** [comment]
This is superseded by 12087, right?

@jszczepkowski @ramr

**Comment by ramr** [comment]
@ncdc  yes that is correct - finishing up the paperwork here. 
Closing in favor of https://github.com/kubernetes/kubernetes/issues/12087

**Pull request closed without merging**